### PR TITLE
Gender selector

### PR DESCRIPTION
**Pull request by edlu77**

### CODE REVIEW

File: projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss [hunk 1]
```diff
@@ -22,7 +22,3 @@
     }
   }
 }
-
-.highlighted {
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: Why are you changing the stage-nav? It is not part of the components you are working on!

File: projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.theme.scss [hunk 2]
```diff
@@ -5,21 +5,12 @@
   $background: map-get($theme, background);
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.theme.scss:5]: This move cannot be correct!? Your components are not part of the stage-nav!

File: projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<div class="organ-sliders">
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.html:1]: This is not what I meant by <ins>**generic component**</ins>! Create a component consisting of a slide-toggle with a label to the right and another label to the left. Then use the <ins>**generic component**</ins> with label inputs `['Left', 'Right']` for one and `['Male', 'Female']` for the other.

File: projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  let shallow: Shallow<LabeledSideToggleComponent>;
+
+  beforeEach(() => {
+    shallow = new Shallow(LabeledSideToggleComponent, LabeledSideToggleModule);
+  });
+
+  it('should emit the correct gender whenever updateGender is called.', async () => {
+    const { instance, outputs } = await shallow.render({ bind: { isMale: true } });
+    instance.updateGender(true);
+    expect(outputs.genderChanged.emit).toHaveBeenCalledWith('female');
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.spec.ts:16]: Both tests should be split into two separate ones. One testing the positive case and one testing the negative case. Having more than one `expect` statement is usually an indicator that it needs to be split.

File: projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss [hunk 5]
```diff
@@ -0,0 +1,18 @@
+@use '~@angular/material/theming' as *;
+
+@mixin theme($theme) {
+  $foreground: map-get($theme, foreground);
+
+  .gender-slider, .side-slider {
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss:6]: Again this NEEDS to be wrapped inside a specific selector i.e. `.ccf-labeled-slide-toggle { ... }`
CHECK other components (block-size-input, rotation-slider, etc.) for how it is done!

File: projects/ccf-rui/src/app/shared/components/organ-selector/organ-selector.component.theme.scss [hunk 6]
```diff
@@ -8,7 +8,7 @@
   .carousel {
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/organ-selector/organ-selector.component.theme.scss:8]: Again this is a completely unrelated component. Why are you changing it in this pr?

File: projects/ccf-rui/src/themes/_light.scss [hunk 7]
```diff
@@ -35,6 +35,8 @@ $foreground: map-merge($mat-light-theme-foreground, (
   slider-green: map-get($mat-green, A700), // #00C853
   slider-red: map-get($mat-red, A700), // #D50000
   text: map-get($primary-colors, 900), // #212121
+  toggle-bar: rgba(25, 118, 210, 0.54),
```
Comment (axdanbol) [inline projects/ccf-rui/src/themes/_light.scss:38]: What are these colors? Don't just move them! Instead use the correct `map-get(...)`. As for the opacity channel it should be applied in the component's theme file using the second for of `mat-color(from, variable, opacity)`

File: projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss (hunk 1 above)
Comment (edlu77) [reply projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: I was working on the fix for the slider color bug and decided to add that to the PR. The only thing I changed was the styling for the view slider.

File: projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss (hunk 5 above)
Comment (edlu77) [reply projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss:6]: Isn't this already selecting the two sliders specifically? Sorry if I'm misunderstanding.

File: projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss (hunk 1 above)
Comment (axdanbol) [reply projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: It is not good practice to add unrelated stuff to the pr as I'm currently working on the stage it may cause conflicts!

Comment (edlu77) [reply projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: Okay sorry, I won't do that again.

Comment (axdanbol) [reply projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: Also if it is a bug it should have its own pr for fixing it! And please link to the issue such as the bug using standard #[issue number] tags

Comment (axdanbol) [reply projects/ccf-rui/src/app/modules/content/stage-nav/stage-nav.component.scss:26]: Well this should be reverted in that case! You can easily revert commits in the  git desktop app by looking at the history tab and right clicking and selecting revert commit on the specific commit.

File: projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss (hunk 5 above)
Comment (axdanbol) [reply projects/ccf-rui/src/app/shared/components/labeled-side-toggle/labeled-side-toggle.component.theme.scss:6]: Yes, but it has the potential of affecting other components. Ex.
Component A theme file:
```
.label {
  color: blue;
}
```

Component B HTML:
```
<span class="label">This text becomes blue because of A</span>
```

Not a problem if component A does this in its theme file:
```
.ccf-component-a { // This now only affects things in component A
  .label {
    color: blue;
  }
}
```

**Comment by bherr2** [comment]
You are missing an l in `ccf-labeled-side-toggle` side => slide.

**Comment by edlu77** [comment]
> You are missing an l in `ccf-labeled-side-toggle` side => slide.

Whoops, I'll change that now.

File: projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ */
+@Component({
+  selector: 'ccf-labeled-slide-toggle',
+  templateUrl: './labeled-slide-toggle.component.html',
+  styleUrls: ['./labeled-slide-toggle.component.scss']
+})
+
+export class LabeledSlideToggleComponent {
+
+  @HostBinding('class') readonly clsName = 'ccf-labeled-slide-toggle';
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:14]: The following comment can be added above variable to make doc coverage happier: `/** HTML class name */`

File: projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * Whether or not the slider is disabled
+   */
+  @Input() slideDisabled = false;
+
+  /**
+   * The category of data selected
+   */
+  @Input() dataType: string;
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:24]: This is not needed as the user of the component already knows this information

File: projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+})
+
+export class LabeledSlideToggleComponent {
+
+  @HostBinding('class') readonly clsName = 'ccf-labeled-slide-toggle';
+
+  /**
+   * Whether or not the slider is disabled
+   */
+  @Input() slideDisabled = false;
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:19]: Call this just `disabled` for consistency with other html components etc.

File: projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * The category of data selected
+   */
+  @Input() dataType: string;
+
+  /**
+   * The two selection options to be toggled
+   */
+  @Input() labels: string[];
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:29]: A better type would be `[string, string]` i.e. a tuple of exactly two strings

File: projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * Determines if left toggle option is selected
+   */
+  left = true;
+
+  /**
+   * Emits the datatype with the currently selected option
+   */
+  @Output() toggleChanged = new EventEmitter<{}>();
```
Comment (axdanbol) [inline projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: Output type will be `string` when the other changes are implemented

Comment (bherr2) [reply projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: This should also ideally be `@Input() value: string` and `@Input() valueChange...` to be consistent with other pure components

Comment (axdanbol) [reply projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: The easiest solution without involving lifecycle hooks would then be something like
```
@Input() set value(label: string) {
  this.left = !this.labels || label !== this.labels[1];
}
```
And if we want to store the value then
```
class LabeledSliderToggle {
  ...

  @Input() set value() ...
  get value(): string { return this._value; }
  private _value: string;
```

Comment (edlu77) [reply projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: What is the valueChange input? I thought the input value was all we needed.

Comment (axdanbol) [reply projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: It is supposed to be `@Output() valueChange ...`

Comment (edlu77) [reply projects/ccf-rui/src/app/shared/components/labeled-slide-toggle/labeled-slide-toggle.component.ts:39]: Ahh ok
